File: engine-python/opptimiz/optimizer.py
# oPPTimiz is a tool that reduces Office documents size.
# Copyright (C) 2025 EDF
# This program is free software: you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later version.
# This program is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
# PARTICULAR PURPOSE. See the GNU General Public License for more details.

"""Cross-platform OOXML (.pptx/.docx/.xlsx) size optimizer.

An OOXML file is a ZIP of XML parts plus media. The document is optimized without
any Office dependency by recompressing the images under */media/, dropping media
that nothing references, and repacking the archive with maximum DEFLATE.
"""

import io
import os
import posixpath
import zipfile
from dataclasses import dataclass

from . import images

LEVELS = {
    "maximal": {"max_dimension": 1920, "jpeg_quality": 72},
    "intermediate": {"max_dimension": 2560, "jpeg_quality": 85},
}

SUPPORTED_SUFFIXES = (".pptx", ".docx", ".xlsx")


@dataclass
class OptimizationResult:
    initial_size: int
    new_size: int
    images_optimized: int
    media_removed: int

    @property
    def gain(self) -> int:
        return self.initial_size - self.new_size

    @property
    def percentage(self) -> float:
        if self.initial_size == 0:
            return 0.0
        return round(self.gain / self.initial_size * 100, 2)


def _is_media(name: str) -> bool:
    return "/media/" in name.lower() and not name.endswith("/")


def _find_unused_media(parts: dict[str, bytes], media_names: list[str]) -> set[str]:
    """Return media parts whose file name appears in no other part.

    Media is referenced from .rels targets (and occasionally inline) by its base
    name. A part is considered removable only when that name is found nowhere
    else; when unsure, the media is kept.
    """
    haystack = bytearray()
    for name, data in parts.items():
        if not _is_media(name):
            haystack += data

    unused = set()
    for name in media_names:
        basename = posixpath.basename(name).encode("utf-8", "ignore")
        if basename and basename not in haystack:
            unused.add(name)
    return unused


def optimize_bytes(
    content: bytes,
    level: str = "maximal",
    prune_unused: bool = True,
) -> tuple[bytes, OptimizationResult]:
    if level not in LEVELS:
        raise ValueError(f"Unknown level '{level}', expected one of {sorted(LEVELS)}")
    preset = LEVELS[level]

    with zipfile.ZipFile(io.BytesIO(content)) as archive:
        order = [info.filename for info in archive.infolist() if not info.is_dir()]
        parts = {name: archive.read(name) for name in order}

    media_names = [name for name in order if _is_media(name)]

    images_optimized = 0
    for name in media_names:
        ext = os.path.splitext(name)[1]
        smaller = images.recompress(parts[name], ext, preset["max_dimension"], preset["jpeg_quality"])
        if smaller is not None:
            parts[name] = smaller
            images_optimized += 1

    media_removed = 0
    if prune_unused:
        for name in _find_unused_media(parts, media_names):
            del parts[name]
            media_removed += 1

    out = io.BytesIO()
    with zipfile.ZipFile(out, "w", zipfile.ZIP_DEFLATED, compresslevel=9) as archive:
        for name in order:
            if name in parts:
                archive.writestr(name, parts[name])

    new_content = out.getvalue()
    result = OptimizationResult(
        initial_size=len(content),
        new_size=len(new_content),
        images_optimized=images_optimized,
        media_removed=media_removed,
    )
    return new_content, result


def optimize_file(
    source: str,
    level: str = "maximal",
    keep_original: bool = False,
    output: str | None = None,
    prune_unused: bool = True,
) -> tuple[str, OptimizationResult]:
    if not source.lower().endswith(SUPPORTED_SUFFIXES):
        raise ValueError(f"Unsupported file type: {source} (expected {SUPPORTED_SUFFIXES})")

    with open(source, "rb") as handle:
        content = handle.read()

    new_content, result = optimize_bytes(content, level=level, prune_unused=prune_unused)

    if output:
        destination = output
    elif keep_original:
        root, ext = os.path.splitext(source)
        destination = f"{root}_oPPTimiz{ext}"
    else:
        destination = source

    with open(destination, "wb") as handle:
        handle.write(new_content)

    return destination, result

File: engine-dotnet/oPPTimiz.Core/Optimizer.cs
//oPPTimiz is a tool that reduces Office documents size.
//Copyright (C) 2025 EDF
//This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
//This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
//You should have received a copy of the GNU General Public License along with this program. If not, see https://www.gnu.org/licenses/.

using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Text;

namespace oPPTimiz.Core
{
    public static class Optimizer
    {
        private static readonly string[] SupportedSuffixes = { ".pptx", ".docx", ".xlsx" };

        /// <summary>
        /// Optimizes an in-memory OOXML document by recompressing images under
        /// */media/, dropping unreferenced media and repacking the archive with
        /// maximum DEFLATE compression.
        /// </summary>
        public static OptimizationResult OptimizeBytes(byte[] content, OptimizationLevel level, bool pruneUnused, out byte[] output)
        {
            int maxDimension = level == OptimizationLevel.Maximal ? 1920 : 2560;
            long jpegQuality = level == OptimizationLevel.Maximal ? 72 : 85;

            var order = new List<string>();
            var parts = new Dictionary<string, byte[]>();

            using (var input = new MemoryStream(content))
            using (var archive = new ZipArchive(input, ZipArchiveMode.Read))
            {
                foreach (var entry in archive.Entries)
                {
                    if (entry.FullName.EndsWith("/"))
                        continue;
                    order.Add(entry.FullName);
                    using (var entryStream = entry.Open())
                    using (var buffer = new MemoryStream())
                    {
                        entryStream.CopyTo(buffer);
                        parts[entry.FullName] = buffer.ToArray();
                    }
                }
            }

            var mediaNames = order.Where(IsMedia).ToList();

            int imagesOptimized = 0;
            foreach (var name in mediaNames)
            {
                byte[] smaller = ImageRecompressor.Recompress(parts[name], Path.GetExtension(name), maxDimension, jpegQuality);
                if (smaller != null)
                {
                    parts[name] = smaller;
                    imagesOptimized++;
                }
            }

            int mediaRemoved = 0;
            if (pruneUnused)
            {
                foreach (var name in FindUnusedMedia(parts, mediaNames))
                {
                    parts.Remove(name);
                    mediaRemoved++;
                }
            }

            byte[] newContent;
            using (var outputStream = new MemoryStream())
            {
                using (var archive = new ZipArchive(outputStream, ZipArchiveMode.Create, true))
                {
                    foreach (var name in order)
                    {
                        if (!parts.ContainsKey(name))
                            continue;
                        var entry = archive.CreateEntry(name, CompressionLevel.Optimal);
                        using (var entryStream = entry.Open())
                        {
                            byte[] data = parts[name];
                            entryStream.Write(data, 0, data.Length);
                        }
                    }
                }
                newContent = outputStream.ToArray();
            }

            output = newContent;
            return new OptimizationResult(content.Length, newContent.Length, imagesOptimized, mediaRemoved);
        }

        public static OptimizationResult OptimizeFile(string source, OptimizationLevel level, bool keepOriginal, string output, out string destination)
        {
            string extension = Path.GetExtension(source).ToLowerInvariant();
            if (!SupportedSuffixes.Contains(extension))
                throw new ArgumentException("Unsupported file type: " + source);

            byte[] content = File.ReadAllBytes(source);

            byte[] newContent;
            OptimizationResult result = OptimizeBytes(content, level, true, out newContent);

            if (!string.IsNullOrEmpty(output))
            {
                destination = output;
            }
            else if (keepOriginal)
            {
                string directory = Path.GetDirectoryName(source);
                string name = Path.GetFileNameWithoutExtension(source);
                destination = Path.Combine(directory, name + "_oPPTimiz" + Path.GetExtension(source));
            }
            else
            {
                destination = source;
            }

            File.WriteAllBytes(destination, newContent);
            return result;
        }

        private static bool IsMedia(string name)
        {
            return name.ToLowerInvariant().Contains("/media/") && !name.EndsWith("/");
        }

        private static IEnumerable<string> FindUnusedMedia(Dictionary<string, byte[]> parts, List<string> mediaNames)
        {
            byte[] haystack;
            using (var buffer = new MemoryStream())
            {
                foreach (var part in parts)
                {
                    if (!IsMedia(part.Key))
                        buffer.Write(part.Value, 0, part.Value.Length);
                }
                haystack = buffer.ToArray();
            }

            var unused = new List<string>();
            foreach (var name in mediaNames)
            {
                byte[] basename = Encoding.UTF8.GetBytes(Path.GetFileName(name));
                if (basename.Length > 0 && ByteIndexOf(haystack, basename) < 0)
                    unused.Add(name);
            }
            return unused;
        }

        private static int ByteIndexOf(byte[] haystack, byte[] needle)
        {
            if (needle.Length == 0 || haystack.Length < needle.Length)
                return -1;

            for (int i = 0; i <= haystack.Length - needle.Length; i++)
            {
                bool match = true;
                for (int j = 0; j < needle.Length; j++)
                {
                    if (haystack[i + j] != needle[j])
                    {
                        match = false;
                        break;
                    }
                }
                if (match)
                    return i;
            }
            return -1;
        }
    }
}
